File: code_snippets/ts/src/develop/journaling_results.ts
import * as restate from "@restatedev/restate-sdk";
import { RestatePromise } from "@restatedev/restate-sdk";
import { MyService } from "./my_service";

const service = restate.service({
  name: "SideEffects",
  handlers: {
    greet: async (ctx: restate.Context, name: string) => {
      // <start_side_effect>
      const result = await ctx.run<string>(async () => doDbRequest());
      // <end_side_effect>
    },

    promiseCombinators: async (ctx: restate.Context, name: string) => {
      // <start_all>
      const sleepPromise = ctx.sleep({ milliseconds: 100 });
      const callPromise = ctx.serviceClient(MyService).myHandler("Hi");
      const externalCallPromise = ctx.run(() => httpCall());
      const resultArray = await RestatePromise.all([
        sleepPromise,
        callPromise,
        externalCallPromise,
      ]);
      // <end_all>

      // <start_uuid>
      const uuid = ctx.rand.uuidv4();
      // <end_uuid>

      // <start_random_nb>
      const randomNumber = ctx.rand.random();
      // <end_random_nb>
    },
  },
});

function doDbRequest() {
  return "";
}

async function httpCall() {
  return "";
}

const paymentClient = {
  call: (txId: string, amt: number) => {
    return { error: "", isSuccess: false };
  },
};

File: code_snippets/go/develop/journalingresults.go
package develop

import (
	"fmt"
	"math/rand/v2"
	"time"

	restate "github.com/restatedev/sdk-go"
)

type JournalingResults struct{}

func (JournalingResults) Greet(ctx restate.Context, name string) error {
	// <start_side_effect>
	result, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return doDbRequest()
	})
	if err != nil {
		return err
	}
	// <end_side_effect>

	_ = result

	return nil
}

func (JournalingResults) Greet2(ctx restate.Context, name string) error {
	// <start_side_effect_retry>
	result, err := restate.Run(ctx,
		func(ctx restate.RunContext) (string, error) {
			return doDbRequest()
		},
		// After 10 seconds, give up retrying
		restate.WithMaxRetryDuration(time.Second*10),
		// On the first retry, wait 100 milliseconds before next attempt
		restate.WithInitialRetryInterval(time.Millisecond*100),
		// Grow retry interval with factor 2
		restate.WithRetryIntervalFactor(2.0),
		// Optional: provide a name for the operation to be visible in the
		// observability tools.
		restate.WithName("my_db_request"),
	)
	if err != nil {
		return err
	}
	// <end_side_effect_retry>

	_ = result

	return nil
}

func (JournalingResults) PromiseCombinators(ctx restate.Context, name string) (string, error) {
	// <start_race>
	sleepFuture := restate.After(ctx, 100*time.Millisecond)
	callFuture := restate.Service[string](ctx, "MyService", "MyHandler").RequestFuture("hi")

	selector := restate.Select(ctx, sleepFuture, callFuture)
	switch selector.Select() {
	case sleepFuture:
		if err := sleepFuture.Done(); err != nil {
			return "", err
		}
		return "sleep won", nil
	case callFuture:
		result, err := callFuture.Response()
		if err != nil {
			return "", err
		}
		return fmt.Sprintf("call won with result: %s", result), nil
	}
	// <end_race>

	// <start_uuid>
	uuid := restate.Rand(ctx).UUID()
	// <end_uuid>

	_ = uuid

	// <start_random_nb>
	randomInt := restate.Rand(ctx).Uint64()
	randomFloat := restate.Rand(ctx).Float64()
	randomSource := rand.New(restate.Rand(ctx).Source())
	// <end_random_nb>

	_ = randomInt
	_ = randomFloat
	_ = randomSource

	return "", nil
}

func doDbRequest() (string, error) {
	return "", nil
}
